# --------------------------------------- Training -------------------------------------

NOTE: This section is for practising. Questions specified by Q are not a part of your coursework. You can check the comments to see what each cell does. 

## Importing necessary Libraries  

In [1]:
import pandas as pd 
import numpy as np

In [2]:
# to download as HTML file 
import plotly.io as pio
pio.renderers.default ='notebook'

In [3]:
# to export as pdf 
!pip install Pyppeteer
!pyppeteer-install

chromium is already installed.


## Python Warmup exercises

* Q1: Write a code that receives an optional text, print all letters in the text and store them in a list

In [4]:
output=[]
text1='''welcome to workshop 1'''
for i in text1:
    print(i)
    output.append(i)
print(output)

w
e
l
c
o
m
e
 
t
o
 
w
o
r
k
s
h
o
p
 
1
['w', 'e', 'l', 'c', 'o', 'm', 'e', ' ', 't', 'o', ' ', 'w', 'o', 'r', 'k', 's', 'h', 'o', 'p', ' ', '1']


* Q2: Write a code that receives an optional text, prints all words in the text and store them in a list

In [5]:
output=[]
text1='''welcome to workshop 1'''
for word in text1.split():
    print(word)
    output.append(word)
print(output)

welcome
to
workshop
1
['welcome', 'to', 'workshop', '1']


* Q3: Write Q2 answer as a function

In [6]:
text1='''welcome to workshop 1'''
def print_word(text):
    for word in text.split():
        print(word)
print_word(text1)

welcome
to
workshop
1


* Q4: Write a program to find out the average of a set of integers. The program prompts the user to enter the count of numbers

In [8]:
count = int(input("Enter the count of numbers: ")) # This gets the count of numbers from the user and 
# puts it in an integer variable
i = 0
sum = 0
for i in range(count):
    x = int(input("Enter an integer: "))
    sum = sum + x
avg = sum/count
print("The average is: ", avg)

Enter the count of numbers: 6
Enter an integer: 9
Enter an integer: 67
Enter an integer: 9
Enter an integer: 55
Enter an integer: 76
Enter an integer: 6
The average is:  37.0


* Q5: Write a program to check whether the given number is even or not. 

In [9]:
# If we divide an even number by 2, the remainder will be zero
number = input("Enter a number: ")
x = int(number)%2
if x == 0:
    print("The number is Even.")
    
else:
    print("The number is Odd.")

Enter a number: 35
The number is Odd.


* Q6: Write a program to find the average of 4 numbers using while loop. The program prompts the user to enter the numbers

In [11]:
count = 0
sum = 0.0
while(count<4):
    number = float(input("Enter a real number: "))
    count=count+1
    sum = sum+number
avg = sum/10
print("Average is :",avg)

Enter a real number: 5
Enter a real number: 78
Enter a real number: 5
Enter a real number: 8
Average is : 9.6


* Q7: Write a function to find the average of a list of integers

In [12]:
def give_me_average(in_list):
    count=len(in_list)
    sum=0
    for i in in_list:
        sum=sum+i
    avg = sum/count
    return avg
a=[1,2,3,4]
print("Average is :",give_me_average(a))    

Average is : 2.5


* Q8: Write a program to display all integers within the range 100-120 whose sum of digits is an even number and store them in a list

In [13]:
output=[]
for i in range(100,120):
    num = i
    sum = 0
    while(num!=0):
        digit = num%10
        sum = sum + digit
        num = num//10
    if(sum%2==0):
        output.append(i)
        print(i)
print(output)

101
103
105
107
109
110
112
114
116
118
[101, 103, 105, 107, 109, 110, 112, 114, 116, 118]


* Q9: Write a program to check whether the given integer is a multiple of both 5 and 7

In [14]:
number = int(input("Enter an integer: "))
if((number%5==0)and(number%7==0)):
    print(number, "is a multiple of both 5 and 7")
else:
    print(number, "is not a multiple of both 5 and 7")

Enter an integer: 305
305 is not a multiple of both 5 and 7


## Pandas warmup exercises

In [15]:
import pandas as pd
# You can use the DataFrame() method to create a Pandas DataFrame from a dictionary, numpy array, or a list.
data = {"calories": [420, 380, 390], "duration": [50, 40, 45]}
#load data into a DataFrame object:
df = pd.DataFrame(data)
print(df)
print('\n') # Addine an empty line to the end of the data frame
# ---
data = [[420, 380, 390],[50, 40, 45]]
df = pd.DataFrame(data)
print(df)


   calories  duration
0       420        50
1       380        40
2       390        45


     0    1    2
0  420  380  390
1   50   40   45


In [16]:
# You can use loc() or iloc() to return values, rows, or columns inside a data frame. Using loc() you need to specify the
# column name, but using iloc() you only need to specify the column index. 
data = {"calories": [420, 380, 390], "duration": [50, 40, 45]}
#load data into a DataFrame object:
df = pd.DataFrame(data)
print(df)
print("\n")
print(df.loc[0,'calories'])
print("\n")
print(df.loc[0,:])
print("\n")
print(df.iloc[:,0])
print("\n")
print(df.iloc[:,1])

   calories  duration
0       420        50
1       380        40
2       390        45


420


calories    420
duration     50
Name: 0, dtype: int64


0    420
1    380
2    390
Name: calories, dtype: int64


0    50
1    40
2    45
Name: duration, dtype: int64


In [17]:
# You can get the indices and columns names using index and column attributes as follows

data = {"calories": [420, 380, 390], "duration": [50, 40, 45]}
#load data into a DataFrame object:
df = pd.DataFrame(data)
index_names=list(df.index)
column_names=list(df.columns)
print(index_names,end='\n\n')
print(column_names)


[0, 1, 2]

['calories', 'duration']


In [18]:
# With the index and columns argument, you can name your own indexes and columns.
data = {
 "calories": [420, 380, 390],
 "duration": [50, 40, 45]
}
df = pd.DataFrame(data, index = ["day1", "day2", "day3"])
print(df)


      calories  duration
day1       420        50
day2       380        40
day3       390        45


In [19]:
# With the index argument, you can name your own indexes and columns
data =[[420, 380, 390],[50, 40, 45]]
df = pd.DataFrame(data, columns = ["c1", "c2", "c3"])
print(df)

    c1   c2   c3
0  420  380  390
1   50   40   45


## Importing a dataset for this workshop

## Importing the dataset directly from the source

In [20]:
# Installing a package which is needed to download the dataset from its online source. This package is recommended by the online source at the following
# URL : https://archive.ics.uci.edu/dataset/2/adult
!pip3 install ucimlrepo

In [21]:
# Downloading the dataset from the online source. The first two lines are given by the online source mentioned above
from ucimlrepo import fetch_ucirepo 
# fetch dataset 
adult = fetch_ucirepo(id=2) 

# Putting data in a pandas dataframe
X = adult.data.features 
y = adult.data.targets 
data=pd.concat([X,y],axis=1)


In [22]:
#printing data
data

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
48837   39           Private  215419  Bachelors             13   
48838   64               NaN  321403    HS-grad              9   
48839   38           Private  374983  Bachelors             13   
48840   44           Private   83891  Bachelors             13   
48841   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation    relationship  \
0           Never-married       Adm-clerical   Not-in-family   
1      Married-civ-spouse    Exec-managerial         Husband   
2                Divorced  Handlers-cleaners   Not-in-family   
3      Married-civ-spouse  Handlers-cleaners         Husband   
4      Married-civ-spouse     Prof-specialty            Wife   
...                   ...                ...             ...   
48837            Divorced     Prof-specialty   Not-in-family   
48838             Widowed                NaN  Other-relative   
48839  Married-civ-spouse     Prof-specialty         Husband   
48840            Divorced       Adm-clerical       Own-child   
48841  Married-civ-spouse    Exec-managerial         Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
2                   White    Male             0             0              40   
3                   Black    Male             0             0              40   
4                   Black  Female             0             0              40   
...                   ...     ...           ...           ...             ...   
48837               White  Female             0             0              36   
48838               Black    Male             0             0              40   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country  income  
0      United-States   <=50K  
1      United-States   <=50K  
2      United-States   <=50K  
3      United-States   <=50K  
4               Cuba   <=50K  
...              ...     ...  
48837  United-States  <=50K.  
48838  United-States  <=50K.  
48839  United-States  <=50K.  
48840  United-States  <=50K.  
48841  United-States   >50K.  

[48842 rows x 15 columns]

## Importing the dataset from a folder on your local disk 

In [23]:
#importing the dataset as a Pandas DataFrame into Python if the dataset is stored on your local hard disk
# You can download the dataset from the following URL: https://archive.ics.uci.edu/dataset/2/adult  
data = pd.read_csv('adult.csv') 
# Replace the current data path with the data path on your local disk 

## Exploring the dataset

In [24]:
# Showing the first 5 rows of the dataset
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [25]:
# Finding the shape of the data
print(data.shape)

(48842, 15)


In [26]:
# Generate a dataset by randomly extracting 30000 rows (samples) 
data_new = data.sample(n=30000, random_state = 48)

In [27]:
# Printing the new dataset
data_new

age  workclass  fnlwgt     education  education-num  \
8029    29    Private  216481       Masters             14   
45203   36    Private  280570  Some-college             10   
27498   25          ?  100903     Bachelors             13   
48416   47    Private  145636     Assoc-voc             11   
43230   33    Private  119422       HS-grad              9   
...    ...        ...     ...           ...            ...   
37544   20    Private  166371       HS-grad              9   
20482   80    Private  202483       HS-grad              9   
39667   20    Private  175808       HS-grad              9   
40001   25  State-gov   31350  Some-college             10   
48523   19    Private  239057       HS-grad              9   

              marital-status         occupation    relationship   race  \
8029      Married-civ-spouse    Exec-managerial            Wife  White   
45203     Married-civ-spouse       Craft-repair         Husband  White   
27498     Married-civ-spouse                  ?            Wife  White   
48416     Married-civ-spouse  Handlers-cleaners         Husband  White   
43230     Married-civ-spouse    Exec-managerial         Husband  White   
...                      ...                ...             ...    ...   
37544          Never-married       Craft-repair  Other-relative  White   
20482  Married-spouse-absent       Adm-clerical   Not-in-family  White   
39667          Never-married       Craft-repair       Own-child  White   
40001          Never-married      Other-service   Not-in-family  White   
48523          Never-married       Craft-repair       Own-child  White   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
8029   Female             0             0              40  United-States   
45203    Male             0             0              45  United-States   
27498  Female             0             0              25  United-States   
48416    Male             0             0              48  United-States   
43230    Male             0             0              40  United-States   
...       ...           ...           ...             ...            ...   
37544    Male             0             0              40  United-States   
20482  Female             0             0              16  United-States   
39667    Male             0             0              40  United-States   
40001    Male             0             0              40  United-States   
48523    Male             0             0              40  United-States   

       income  
8029     >50K  
45203  <=50K.  
27498   <=50K  
48416   >50K.  
43230  <=50K.  
...       ...  
37544  <=50K.  
20482   <=50K  
39667  <=50K.  
40001  <=50K.  
48523  <=50K.  

[30000 rows x 15 columns]

In [28]:
# The indices of different rows in the dataset are currently messy. This happens in many data science projects. Always reindex 
# the dataset if you are unsure the indices are correct.  
data_new.reset_index(drop=True, inplace=True)

In [29]:
# Checking if the indices are correct
data_new

age  workclass  fnlwgt     education  education-num  \
0       29    Private  216481       Masters             14   
1       36    Private  280570  Some-college             10   
2       25          ?  100903     Bachelors             13   
3       47    Private  145636     Assoc-voc             11   
4       33    Private  119422       HS-grad              9   
...    ...        ...     ...           ...            ...   
29995   20    Private  166371       HS-grad              9   
29996   80    Private  202483       HS-grad              9   
29997   20    Private  175808       HS-grad              9   
29998   25  State-gov   31350  Some-college             10   
29999   19    Private  239057       HS-grad              9   

              marital-status         occupation    relationship   race  \
0         Married-civ-spouse    Exec-managerial            Wife  White   
1         Married-civ-spouse       Craft-repair         Husband  White   
2         Married-civ-spouse                  ?            Wife  White   
3         Married-civ-spouse  Handlers-cleaners         Husband  White   
4         Married-civ-spouse    Exec-managerial         Husband  White   
...                      ...                ...             ...    ...   
29995          Never-married       Craft-repair  Other-relative  White   
29996  Married-spouse-absent       Adm-clerical   Not-in-family  White   
29997          Never-married       Craft-repair       Own-child  White   
29998          Never-married      Other-service   Not-in-family  White   
29999          Never-married       Craft-repair       Own-child  White   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
0      Female             0             0              40  United-States   
1        Male             0             0              45  United-States   
2      Female             0             0              25  United-States   
3        Male             0             0              48  United-States   
4        Male             0             0              40  United-States   
...       ...           ...           ...             ...            ...   
29995    Male             0             0              40  United-States   
29996  Female             0             0              16  United-States   
29997    Male             0             0              40  United-States   
29998    Male             0             0              40  United-States   
29999    Male             0             0              40  United-States   

       income  
0        >50K  
1      <=50K.  
2       <=50K  
3       >50K.  
4      <=50K.  
...       ...  
29995  <=50K.  
29996   <=50K  
29997  <=50K.  
29998  <=50K.  
29999  <=50K.  

[30000 rows x 15 columns]

In [30]:
data

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
48837   39           Private  215419  Bachelors             13   
48838   64               NaN  321403    HS-grad              9   
48839   38           Private  374983  Bachelors             13   
48840   44           Private   83891  Bachelors             13   
48841   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation    relationship  \
0           Never-married       Adm-clerical   Not-in-family   
1      Married-civ-spouse    Exec-managerial         Husband   
2                Divorced  Handlers-cleaners   Not-in-family   
3      Married-civ-spouse  Handlers-cleaners         Husband   
4      Married-civ-spouse     Prof-specialty            Wife   
...                   ...                ...             ...   
48837            Divorced     Prof-specialty   Not-in-family   
48838             Widowed                NaN  Other-relative   
48839  Married-civ-spouse     Prof-specialty         Husband   
48840            Divorced       Adm-clerical       Own-child   
48841  Married-civ-spouse    Exec-managerial         Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
2                   White    Male             0             0              40   
3                   Black    Male             0             0              40   
4                   Black  Female             0             0              40   
...                   ...     ...           ...           ...             ...   
48837               White  Female             0             0              36   
48838               Black    Male             0             0              40   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country  income  
0      United-States   <=50K  
1      United-States   <=50K  
2      United-States   <=50K  
3      United-States   <=50K  
4               Cuba   <=50K  
...              ...     ...  
48837  United-States  <=50K.  
48838  United-States  <=50K.  
48839  United-States  <=50K.  
48840  United-States  <=50K.  
48841  United-States   >50K.  

[48842 rows x 15 columns]

In [31]:
# Getting statistical information of the dataset for different columns (features) 
data.describe(include="all")

age workclass        fnlwgt education  education-num  \
count   48842.000000     47879  4.884200e+04     48842   48842.000000   
unique           NaN         9           NaN        16            NaN   
top              NaN   Private           NaN   HS-grad            NaN   
freq             NaN     33906           NaN     15784            NaN   
mean       38.643585       NaN  1.896641e+05       NaN      10.078089   
std        13.710510       NaN  1.056040e+05       NaN       2.570973   
min        17.000000       NaN  1.228500e+04       NaN       1.000000   
25%        28.000000       NaN  1.175505e+05       NaN       9.000000   
50%        37.000000       NaN  1.781445e+05       NaN      10.000000   
75%        48.000000       NaN  2.376420e+05       NaN      12.000000   
max        90.000000       NaN  1.490400e+06       NaN      16.000000   

            marital-status      occupation relationship   race    sex  \
count                48842           47876        48842  48842  48842   
unique                   7              15            6      5      2   
top     Married-civ-spouse  Prof-specialty      Husband  White   Male   
freq                 22379            6172        19716  41762  32650   
mean                   NaN             NaN          NaN    NaN    NaN   
std                    NaN             NaN          NaN    NaN    NaN   
min                    NaN             NaN          NaN    NaN    NaN   
25%                    NaN             NaN          NaN    NaN    NaN   
50%                    NaN             NaN          NaN    NaN    NaN   
75%                    NaN             NaN          NaN    NaN    NaN   
max                    NaN             NaN          NaN    NaN    NaN   

        capital-gain  capital-loss  hours-per-week native-country income  
count   48842.000000  48842.000000    48842.000000          48568  48842  
unique           NaN           NaN             NaN             42      4  
top              NaN           NaN             NaN  United-States  <=50K  
freq             NaN           NaN             NaN          43832  24720  
mean     1079.067626     87.502314       40.422382            NaN    NaN  
std      7452.019058    403.004552       12.391444            NaN    NaN  
min         0.000000      0.000000        1.000000            NaN    NaN  
25%         0.000000      0.000000       40.000000            NaN    NaN  
50%         0.000000      0.000000       40.000000            NaN    NaN  
75%         0.000000      0.000000       45.000000            NaN    NaN  
max     99999.000000   4356.000000       99.000000            NaN    NaN

In [32]:
# Droping row with Null values (This topic will be discussed in detail in the next lecture)
data.dropna(inplace=True)

In [33]:
data

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
48836   33           Private  245211  Bachelors             13   
48837   39           Private  215419  Bachelors             13   
48839   38           Private  374983  Bachelors             13   
48840   44           Private   83891  Bachelors             13   
48841   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation   relationship  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
...                   ...                ...            ...   
48836       Never-married     Prof-specialty      Own-child   
48837            Divorced     Prof-specialty  Not-in-family   
48839  Married-civ-spouse     Prof-specialty        Husband   
48840            Divorced       Adm-clerical      Own-child   
48841  Married-civ-spouse    Exec-managerial        Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
2                   White    Male             0             0              40   
3                   Black    Male             0             0              40   
4                   Black  Female             0             0              40   
...                   ...     ...           ...           ...             ...   
48836               White    Male             0             0              40   
48837               White  Female             0             0              36   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country  income  
0      United-States   <=50K  
1      United-States   <=50K  
2      United-States   <=50K  
3      United-States   <=50K  
4               Cuba   <=50K  
...              ...     ...  
48836  United-States  <=50K.  
48837  United-States  <=50K.  
48839  United-States  <=50K.  
48840  United-States  <=50K.  
48841  United-States   >50K.  

[47621 rows x 15 columns]

In [34]:
# reindexing the original data
data.reset_index(drop=True, inplace=True)

In [35]:
data

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
47616   33           Private  245211  Bachelors             13   
47617   39           Private  215419  Bachelors             13   
47618   38           Private  374983  Bachelors             13   
47619   44           Private   83891  Bachelors             13   
47620   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation   relationship  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
...                   ...                ...            ...   
47616       Never-married     Prof-specialty      Own-child   
47617            Divorced     Prof-specialty  Not-in-family   
47618  Married-civ-spouse     Prof-specialty        Husband   
47619            Divorced       Adm-clerical      Own-child   
47620  Married-civ-spouse    Exec-managerial        Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
2                   White    Male             0             0              40   
3                   Black    Male             0             0              40   
4                   Black  Female             0             0              40   
...                   ...     ...           ...           ...             ...   
47616               White    Male             0             0              40   
47617               White  Female             0             0              36   
47618               White    Male             0             0              50   
47619  Asian-Pac-Islander    Male          5455             0              40   
47620               White    Male             0             0              60   

      native-country  income  
0      United-States   <=50K  
1      United-States   <=50K  
2      United-States   <=50K  
3      United-States   <=50K  
4               Cuba   <=50K  
...              ...     ...  
47616  United-States  <=50K.  
47617  United-States  <=50K.  
47618  United-States  <=50K.  
47619  United-States  <=50K.  
47620  United-States   >50K.  

[47621 rows x 15 columns]

In [36]:
# Getting the number of Null values for different columns
data.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [37]:
# Showing the dataset information
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47621 entries, 0 to 47620
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             47621 non-null  int64 
 1   workclass       47621 non-null  object
 2   fnlwgt          47621 non-null  int64 
 3   education       47621 non-null  object
 4   education-num   47621 non-null  int64 
 5   marital-status  47621 non-null  object
 6   occupation      47621 non-null  object
 7   relationship    47621 non-null  object
 8   race            47621 non-null  object
 9   sex             47621 non-null  object
 10  capital-gain    47621 non-null  int64 
 11  capital-loss    47621 non-null  int64 
 12  hours-per-week  47621 non-null  int64 
 13  native-country  47621 non-null  object
 14  income          47621 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.4+ MB


In [38]:
data.shape

(47621, 15)

In [39]:
# Getting the count of different values in the column "education-num"
data['education-num'].value_counts()

education-num
9     15444
10    10512
13     7881
14     2610
11     2034
7      1746
12     1566
6      1336
4       912
15      819
5       735
8       633
16      582
3       494
2       239
1        78
Name: count, dtype: int64

In [40]:
# Getting the count of different values in the column "education"
data['education'].value_counts()

education
HS-grad         15444
Some-college    10512
Bachelors        7881
Masters          2610
Assoc-voc        2034
11th             1746
Assoc-acdm       1566
10th             1336
7th-8th           912
Prof-school       819
9th               735
12th              633
Doctorate         582
5th-6th           494
1st-4th           239
Preschool          78
Name: count, dtype: int64

In [41]:
# Dropping a column
data = data.drop(['fnlwgt'], axis=1)

In [42]:
data.shape

(47621, 14)

In [43]:
# Getting the number of unique values of a column
data['education'].nunique()

16

In [44]:
# Finding how many rows are related to either gender 
data['sex'].value_counts()

sex
Male      31937
Female    15684
Name: count, dtype: int64

In [45]:
# Calculating the average age of different genders in the dataset 
data['age'].groupby([data['sex']]).mean()


sex
Female    36.961043
Male      39.465542
Name: age, dtype: float64

In [46]:
# Getting the average age of different genders in the dataset broken down based on their education
data['age'].groupby([data['sex'],data['education']]).mean()

sex     education   
Female  10th            36.520642
        11th            29.891410
        12th            29.716418
        1st-4th         46.508475
        5th-6th         45.346774
        7th-8th         51.088889
        9th             41.566038
        Assoc-acdm      36.402619
        Assoc-voc       38.000000
        Bachelors       35.747837
        Doctorate       45.557522
        HS-grad         38.862285
        Masters         42.520482
        Preschool       44.409091
        Prof-school     40.131783
        Some-college    33.778719
Male    10th            38.406667
        11th            33.479185
        12th            33.842593
        1st-4th         46.050000
        5th-6th         42.524324
        7th-8th         48.608443
        9th             40.556405
        Assoc-acdm      38.537173
        Assoc-voc       38.853881
        Bachelors       40.190319
        Doctorate       47.936034
        HS-grad         39.073626
        Masters         44.

In [47]:
# Getting the maximum age of different races in the dataset
data['age'].groupby([data['race']]).max() 

race
Amer-Indian-Eskimo    82
Asian-Pac-Islander    90
Black                 90
Other                 77
White                 90
Name: age, dtype: int64

In [48]:
data

age         workclass  education  education-num      marital-status  \
0       39         State-gov  Bachelors             13       Never-married   
1       50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   
2       38           Private    HS-grad              9            Divorced   
3       53           Private       11th              7  Married-civ-spouse   
4       28           Private  Bachelors             13  Married-civ-spouse   
...    ...               ...        ...            ...                 ...   
47616   33           Private  Bachelors             13       Never-married   
47617   39           Private  Bachelors             13            Divorced   
47618   38           Private  Bachelors             13  Married-civ-spouse   
47619   44           Private  Bachelors             13            Divorced   
47620   35      Self-emp-inc  Bachelors             13  Married-civ-spouse   

              occupation   relationship                race     sex  \
0           Adm-clerical  Not-in-family               White    Male   
1        Exec-managerial        Husband               White    Male   
2      Handlers-cleaners  Not-in-family               White    Male   
3      Handlers-cleaners        Husband               Black    Male   
4         Prof-specialty           Wife               Black  Female   
...                  ...            ...                 ...     ...   
47616     Prof-specialty      Own-child               White    Male   
47617     Prof-specialty  Not-in-family               White  Female   
47618     Prof-specialty        Husband               White    Male   
47619       Adm-clerical      Own-child  Asian-Pac-Islander    Male   
47620    Exec-managerial        Husband               White    Male   

       capital-gain  capital-loss  hours-per-week native-country  income  
0              2174             0              40  United-States   <=50K  
1                 0             0              13  United-States   <=50K  
2                 0             0              40  United-States   <=50K  
3                 0             0              40  United-States   <=50K  
4                 0             0              40           Cuba   <=50K  
...             ...           ...             ...            ...     ...  
47616             0             0              40  United-States  <=50K.  
47617             0             0              36  United-States  <=50K.  
47618             0             0              50  United-States  <=50K.  
47619          5455             0              40  United-States  <=50K.  
47620             0             0              60  United-States   >50K.  

[47621 rows x 14 columns]

In [49]:
# Extracting the age and education columns and creating a new DataFrame using these columns
a=data['age']
b=data['education']
new_data=pd.concat([a,b],axis=1)

In [50]:
new_data

age  education
0       39  Bachelors
1       50  Bachelors
2       38    HS-grad
3       53       11th
4       28  Bachelors
...    ...        ...
47616   33  Bachelors
47617   39  Bachelors
47618   38  Bachelors
47619   44  Bachelors
47620   35  Bachelors

[47621 rows x 2 columns]

* Q10: Write a function that receives the dataset and replace Famle with F and Male with M (please try to write it yourself before checking the answer in the next cell)

In [51]:
# Answer
def encode_sex(data):
    rows=data.shape[0]
    a=data.loc[:,'sex']
    for i in range(rows): 
        if a[i]=="Male":
            data.loc[i,"sex"]="M"
        elif a[i]=="Female":
            data.loc[i,"sex"]="F"
    return data

In [52]:
# Copying the data
data_copy=data.copy()

In [53]:
# Applying the encode_sex function to the copied data
data_encoded=encode_sex(data_copy)
data_encoded.head()

age         workclass  education  education-num      marital-status  \
0   39         State-gov  Bachelors             13       Never-married   
1   50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   
2   38           Private    HS-grad              9            Divorced   
3   53           Private       11th              7  Married-civ-spouse   
4   28           Private  Bachelors             13  Married-civ-spouse   

          occupation   relationship   race sex  capital-gain  capital-loss  \
0       Adm-clerical  Not-in-family  White   M          2174             0   
1    Exec-managerial        Husband  White   M             0             0   
2  Handlers-cleaners  Not-in-family  White   M             0             0   
3  Handlers-cleaners        Husband  Black   M             0             0   
4     Prof-specialty           Wife  Black   F             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              13  United-States  <=50K  
2              40  United-States  <=50K  
3              40  United-States  <=50K  
4              40           Cuba  <=50K

In [54]:
data

age         workclass  education  education-num      marital-status  \
0       39         State-gov  Bachelors             13       Never-married   
1       50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   
2       38           Private    HS-grad              9            Divorced   
3       53           Private       11th              7  Married-civ-spouse   
4       28           Private  Bachelors             13  Married-civ-spouse   
...    ...               ...        ...            ...                 ...   
47616   33           Private  Bachelors             13       Never-married   
47617   39           Private  Bachelors             13            Divorced   
47618   38           Private  Bachelors             13  Married-civ-spouse   
47619   44           Private  Bachelors             13            Divorced   
47620   35      Self-emp-inc  Bachelors             13  Married-civ-spouse   

              occupation   relationship                race     sex  \
0           Adm-clerical  Not-in-family               White    Male   
1        Exec-managerial        Husband               White    Male   
2      Handlers-cleaners  Not-in-family               White    Male   
3      Handlers-cleaners        Husband               Black    Male   
4         Prof-specialty           Wife               Black  Female   
...                  ...            ...                 ...     ...   
47616     Prof-specialty      Own-child               White    Male   
47617     Prof-specialty  Not-in-family               White  Female   
47618     Prof-specialty        Husband               White    Male   
47619       Adm-clerical      Own-child  Asian-Pac-Islander    Male   
47620    Exec-managerial        Husband               White    Male   

       capital-gain  capital-loss  hours-per-week native-country  income  
0              2174             0              40  United-States   <=50K  
1                 0             0              13  United-States   <=50K  
2                 0             0              40  United-States   <=50K  
3                 0             0              40  United-States   <=50K  
4                 0             0              40           Cuba   <=50K  
...             ...           ...             ...            ...     ...  
47616             0             0              40  United-States  <=50K.  
47617             0             0              36  United-States  <=50K.  
47618             0             0              50  United-States  <=50K.  
47619          5455             0              40  United-States  <=50K.  
47620             0             0              60  United-States   >50K.  

[47621 rows x 14 columns]

# Tasks

## TASK 1.1: Import the dataset from the URL we used in this workshop. Then generate a new dataset by randomly extracting 10000 samples. Reindex the generated dataset and remove NULL values. Name the new dataset 'task_dataset'

In [55]:
from ucimlrepo import fetch_ucirepo 
# fetch dataset 
adult = fetch_ucirepo(id=2) 

# Putting data in a pandas dataframe
X = adult.data.features 
y = adult.data.targets 
data = pd.concat([X,y],axis=1)

In [56]:
############# WRITE THE CODE IN THIS CELL ####################

# Generate a dataset by randomly extracting 30000 rows (samples) 
task_dataset = data.sample(n=10000, random_state = 48)

In [57]:
task_dataset

age         workclass  fnlwgt     education  education-num  \
8029    29           Private  216481       Masters             14   
45203   36           Private  280570  Some-college             10   
27498   25                 ?  100903     Bachelors             13   
48416   47           Private  145636     Assoc-voc             11   
43230   33           Private  119422       HS-grad              9   
...    ...               ...     ...           ...            ...   
5637    19           Private   63363  Some-college             10   
46277   53           Private   58535       HS-grad              9   
26200   30           Private  342709       HS-grad              9   
33519   41  Self-emp-not-inc  134724     Assoc-voc             11   
28113   21           Private  252253  Some-college             10   

           marital-status         occupation   relationship   race     sex  \
8029   Married-civ-spouse    Exec-managerial           Wife  White  Female   
45203  Married-civ-spouse       Craft-repair        Husband  White    Male   
27498  Married-civ-spouse                  ?           Wife  White  Female   
48416  Married-civ-spouse  Handlers-cleaners        Husband  White    Male   
43230  Married-civ-spouse    Exec-managerial        Husband  White    Male   
...                   ...                ...            ...    ...     ...   
5637        Never-married              Sales      Own-child  White  Female   
46277            Divorced              Sales  Not-in-family  White  Female   
26200  Married-civ-spouse  Handlers-cleaners        Husband  White    Male   
33519  Married-civ-spouse      Other-service           Wife  White  Female   
28113       Never-married       Adm-clerical      Unmarried  Black  Female   

       capital-gain  capital-loss  hours-per-week native-country  income  
8029              0             0              40  United-States    >50K  
45203             0             0              45  United-States  <=50K.  
27498             0             0              25  United-States   <=50K  
48416             0             0              48  United-States   >50K.  
43230             0             0              40  United-States  <=50K.  
...             ...           ...             ...            ...     ...  
5637              0             0              30  United-States   <=50K  
46277             0             0              40  United-States  <=50K.  
26200             0             0              40  United-States   <=50K  
33519          3103             0              40  United-States   >50K.  
28113             0             0              40  United-States   <=50K  

[10000 rows x 15 columns]

In [58]:
# The indices of different rows from the randomly generated dataset are currently messy
# reindex randomly generated dataset
task_dataset.reset_index(drop=True, inplace=True)

In [59]:
# Checking if the indices are correct
task_dataset

age         workclass  fnlwgt     education  education-num  \
0      29           Private  216481       Masters             14   
1      36           Private  280570  Some-college             10   
2      25                 ?  100903     Bachelors             13   
3      47           Private  145636     Assoc-voc             11   
4      33           Private  119422       HS-grad              9   
...   ...               ...     ...           ...            ...   
9995   19           Private   63363  Some-college             10   
9996   53           Private   58535       HS-grad              9   
9997   30           Private  342709       HS-grad              9   
9998   41  Self-emp-not-inc  134724     Assoc-voc             11   
9999   21           Private  252253  Some-college             10   

          marital-status         occupation   relationship   race     sex  \
0     Married-civ-spouse    Exec-managerial           Wife  White  Female   
1     Married-civ-spouse       Craft-repair        Husband  White    Male   
2     Married-civ-spouse                  ?           Wife  White  Female   
3     Married-civ-spouse  Handlers-cleaners        Husband  White    Male   
4     Married-civ-spouse    Exec-managerial        Husband  White    Male   
...                  ...                ...            ...    ...     ...   
9995       Never-married              Sales      Own-child  White  Female   
9996            Divorced              Sales  Not-in-family  White  Female   
9997  Married-civ-spouse  Handlers-cleaners        Husband  White    Male   
9998  Married-civ-spouse      Other-service           Wife  White  Female   
9999       Never-married       Adm-clerical      Unmarried  Black  Female   

      capital-gain  capital-loss  hours-per-week native-country  income  
0                0             0              40  United-States    >50K  
1                0             0              45  United-States  <=50K.  
2                0             0              25  United-States   <=50K  
3                0             0              48  United-States   >50K.  
4                0             0              40  United-States  <=50K.  
...            ...           ...             ...            ...     ...  
9995             0             0              30  United-States   <=50K  
9996             0             0              40  United-States  <=50K.  
9997             0             0              40  United-States   <=50K  
9998          3103             0              40  United-States   >50K.  
9999             0             0              40  United-States   <=50K  

[10000 rows x 15 columns]

In [60]:
# Getting the number of Null values for different columns
task_dataset.isna().sum()

age                 0
workclass         175
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        175
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country     61
income              0
dtype: int64

We have identified workclass,occupation and native-country have missing values

In [61]:
# Droping row with Null values
task_dataset.dropna(inplace=True)

In [62]:
# Getting the number of Null values for different columns
task_dataset.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [63]:
task_dataset

age         workclass  fnlwgt     education  education-num  \
0      29           Private  216481       Masters             14   
1      36           Private  280570  Some-college             10   
2      25                 ?  100903     Bachelors             13   
3      47           Private  145636     Assoc-voc             11   
4      33           Private  119422       HS-grad              9   
...   ...               ...     ...           ...            ...   
9995   19           Private   63363  Some-college             10   
9996   53           Private   58535       HS-grad              9   
9997   30           Private  342709       HS-grad              9   
9998   41  Self-emp-not-inc  134724     Assoc-voc             11   
9999   21           Private  252253  Some-college             10   

          marital-status         occupation   relationship   race     sex  \
0     Married-civ-spouse    Exec-managerial           Wife  White  Female   
1     Married-civ-spouse       Craft-repair        Husband  White    Male   
2     Married-civ-spouse                  ?           Wife  White  Female   
3     Married-civ-spouse  Handlers-cleaners        Husband  White    Male   
4     Married-civ-spouse    Exec-managerial        Husband  White    Male   
...                  ...                ...            ...    ...     ...   
9995       Never-married              Sales      Own-child  White  Female   
9996            Divorced              Sales  Not-in-family  White  Female   
9997  Married-civ-spouse  Handlers-cleaners        Husband  White    Male   
9998  Married-civ-spouse      Other-service           Wife  White  Female   
9999       Never-married       Adm-clerical      Unmarried  Black  Female   

      capital-gain  capital-loss  hours-per-week native-country  income  
0                0             0              40  United-States    >50K  
1                0             0              45  United-States  <=50K.  
2                0             0              25  United-States   <=50K  
3                0             0              48  United-States   >50K.  
4                0             0              40  United-States  <=50K.  
...            ...           ...             ...            ...     ...  
9995             0             0              30  United-States   <=50K  
9996             0             0              40  United-States  <=50K.  
9997             0             0              40  United-States   <=50K  
9998          3103             0              40  United-States   >50K.  
9999             0             0              40  United-States   <=50K  

[9765 rows x 15 columns]

In [64]:
task_dataset.head()

age workclass  fnlwgt     education  education-num      marital-status  \
0   29   Private  216481       Masters             14  Married-civ-spouse   
1   36   Private  280570  Some-college             10  Married-civ-spouse   
2   25         ?  100903     Bachelors             13  Married-civ-spouse   
3   47   Private  145636     Assoc-voc             11  Married-civ-spouse   
4   33   Private  119422       HS-grad              9  Married-civ-spouse   

          occupation relationship   race     sex  capital-gain  capital-loss  \
0    Exec-managerial         Wife  White  Female             0             0   
1       Craft-repair      Husband  White    Male             0             0   
2                  ?         Wife  White  Female             0             0   
3  Handlers-cleaners      Husband  White    Male             0             0   
4    Exec-managerial      Husband  White    Male             0             0   

   hours-per-week native-country  income  
0              40  United-States    >50K  
1              45  United-States  <=50K.  
2              25  United-States   <=50K  
3              48  United-States   >50K.  
4              40  United-States  <=50K.


A new dataset named 'task_dataset' was randomly generated from the 'Adult' dataset imported from 'fetch_ucirepo'.The random dataset reindexed and dropped null values.

## TASK 1.2: Write a code to find how much contribution each sex and occupation category made to the capital-gain on average. Apply the code to the task_dataset and report the result (Hint: you need to use the groupby method)

In [65]:
#display columns in dataset
task_dataset.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='object')

In [66]:
# Getting the average capital-gain of different genders in the dataset broken down based on their occupation
task_dataset.groupby(['sex', 'occupation'])['capital-gain'].mean().reset_index()

sex         occupation  capital-gain
0   Female                  ?    359.855556
1   Female       Adm-clerical    508.217678
2   Female       Craft-repair    273.969231
3   Female    Exec-managerial   1311.540984
4   Female    Farming-fishing   1049.523810
5   Female  Handlers-cleaners    199.136364
6   Female  Machine-op-inspct    290.217687
7   Female      Other-service    191.525394
8   Female    Priv-house-serv     74.425926
9   Female     Prof-specialty   1008.057269
10  Female    Protective-serv   5081.000000
11  Female              Sales    276.654822
12  Female       Tech-support    382.803419
13  Female   Transport-moving    533.800000
14    Male                  ?    879.403509
15    Male       Adm-clerical    425.640751
16    Male       Armed-Forces      0.000000
17    Male       Craft-repair    906.637306
18    Male    Exec-managerial   2339.700234
19    Male    Farming-fishing    441.013746
20    Male  Handlers-cleaners    149.265957
21    Male  Machine-op-inspct    577.642082
22    Male      Other-service    155.411765
23    Male    Priv-house-serv    118.800000
24    Male     Prof-specialty   3918.730198
25    Male    Protective-serv    368.563158
26    Male              Sales   1129.695105
27    Male       Tech-support    419.494565
28    Male   Transport-moving    409.625000

## TASK 1.3: Write a code to find the country with the highest number of people with a Bachelors degree. Apply the code to the task_dataset and report the result (Hint: you need to use the groupby method)

In [67]:
# Filter Bachelors degree values from education
Bachelors_Degree = task_dataset[task_dataset['education'] == 'Bachelors']

# group by country and count number of occurences 
bsc_country_count = Bachelors_Degree.groupby('native-country').size().reset_index(name='count')

In [68]:
bsc_country_count

native-country  count
0                    ?     30
1             Cambodia      2
2               Canada      7
3                China      6
4             Columbia      1
5                 Cuba      5
6   Dominican-Republic      2
7              Ecuador      1
8          El-Salvador      4
9              England      9
10             Germany      5
11              Greece      1
12           Guatemala      1
13               India      7
14                Iran      4
15             Ireland      1
16               Italy      6
17             Jamaica      3
18               Japan      8
19              Mexico      7
20                Peru      2
21         Philippines     15
22              Poland      1
23         Puerto-Rico      6
24               South      6
25              Taiwan      5
26            Thailand      2
27       United-States   1456
28             Vietnam      3

In [69]:
# Find the country with the highest number of people with a Bachelors degree
country_highest_bachelors = bsc_country_count['count'].max()
country_highest_bachelors = bsc_country_count[bsc_country_count['count'] == country_highest_bachelors]['native-country'].iloc[0]

In [70]:
# Print the country with the highest number of people with a Bachelors degree
print(f"Country with the highest number of people with a Bachelors degree: {country_highest_bachelors}")

Country with the highest number of people with a Bachelors degree: United-States


The dataset was filtered to include only the rows where the 'education' column is equal to 'Bachelors' and the number of individuals with a Bachelor's degree from each country.
The task_dataset was further analyzed to determine which country has the highest number of people with a Bachelor's degree with the result being the United States.

## TASK 1.4: Write a code to receive two lists including five names and their respective ages and print 'Hello Name Age' 
## For example, if it received a list of two names ['Amin', 'Michael'] and respective ages [27,38], it would print 'Hello Amin 27', 'Hello Michael 38'. Each hello statement should be printed in a new line 

In [71]:
names = ['Opeyemi', 'Sam', 'Lekan', 'Victory', 'Omohjesu']
ages = [27, 38, 37, 30, 21]

# iterate both list
for name, age in zip(names, ages):
    
    # print desired result
    print(f'Hello {name} {age}')

Hello Opeyemi 27
Hello Sam 38
Hello Lekan 37
Hello Victory 30
Hello Omohjesu 21


Created two list 'Name and Ages' with five values each and looped the corresponding name and age f-string formatting to print desired result

## TASK 1.5: Write a code to receive an optional text, capitalise all words in the text and print them

In [72]:
def capitalize_words(text=None):
    if text:
        capitalized_text = text.upper()
        print(capitalized_text)
        
input_text = "Jesus is Lord"
capitalize_words(input_text)

JESUS IS LORD


## TASK 1.6: Write a function to split the task_dataset in half column-wise and swap the first half and the second half

In [73]:
def swap_columns(task_dataset):
    # Split the dataset in half colunms
    num_colunms = len(task_dataset.columns)
    half_point = num_colunms // 2
    first_half = task_dataset.iloc[:, :half_point]
    second_half = task_dataset.iloc[:, half_point:]
    
    #swap the first and second halves
    swapped_task_dataset = pd.concat([second_half, first_half], axis=1)
    
    return swapped_task_dataset
    
swap_columns(task_dataset)

relationship   race     sex  capital-gain  capital-loss  \
0              Wife  White  Female             0             0   
1           Husband  White    Male             0             0   
2              Wife  White  Female             0             0   
3           Husband  White    Male             0             0   
4           Husband  White    Male             0             0   
...             ...    ...     ...           ...           ...   
9995      Own-child  White  Female             0             0   
9996  Not-in-family  White  Female             0             0   
9997        Husband  White    Male             0             0   
9998           Wife  White  Female          3103             0   
9999      Unmarried  Black  Female             0             0   

      hours-per-week native-country  income  age         workclass  fnlwgt  \
0                 40  United-States    >50K   29           Private  216481   
1                 45  United-States  <=50K.   36           Private  280570   
2                 25  United-States   <=50K   25                 ?  100903   
3                 48  United-States   >50K.   47           Private  145636   
4                 40  United-States  <=50K.   33           Private  119422   
...              ...            ...     ...  ...               ...     ...   
9995              30  United-States   <=50K   19           Private   63363   
9996              40  United-States  <=50K.   53           Private   58535   
9997              40  United-States   <=50K   30           Private  342709   
9998              40  United-States   >50K.   41  Self-emp-not-inc  134724   
9999              40  United-States   <=50K   21           Private  252253   

         education  education-num      marital-status         occupation  
0          Masters             14  Married-civ-spouse    Exec-managerial  
1     Some-college             10  Married-civ-spouse       Craft-repair  
2        Bachelors             13  Married-civ-spouse                  ?  
3        Assoc-voc             11  Married-civ-spouse  Handlers-cleaners  
4          HS-grad              9  Married-civ-spouse    Exec-managerial  
...            ...            ...                 ...                ...  
9995  Some-college             10       Never-married              Sales  
9996       HS-grad              9            Divorced              Sales  
9997       HS-grad              9  Married-civ-spouse  Handlers-cleaners  
9998     Assoc-voc             11  Married-civ-spouse      Other-service  
9999  Some-college             10       Never-married       Adm-clerical  

[9765 rows x 15 columns]

We divided the columns in the dataset into half and swapped positions of DataFrame halves by concatenating along columns axis.Relationship - income columns are now before age - occupation 

## Task 1.7: Write a function that receives two numerical columns' names and compare their values for all rows. If the value of the first column is greater than the second column, the function should produce True, otherwise, it should produce False. The function should append an additional column to the dataset to store the results of the comparison for all rows. Apply the function to the "age" and "hours-per-week" columns in the task_dataset and print the result.

In [74]:
# Compare values for each row and create a new column with the results
def compare_colunms(task_dataset, col1, col2):
    
    #perform comparison and create new column with the results
    task_dataset['comparison_result'] = task_dataset[col1] > task_dataset[col2] 

    return task_dataset

# Apply the function to task_dataset
result_task_dataset = compare_colunms(task_dataset, 'age', 'hours-per-week')

# print the result
print(result_task_dataset)



      age         workclass  fnlwgt     education  education-num  \
0      29           Private  216481       Masters             14   
1      36           Private  280570  Some-college             10   
2      25                 ?  100903     Bachelors             13   
3      47           Private  145636     Assoc-voc             11   
4      33           Private  119422       HS-grad              9   
...   ...               ...     ...           ...            ...   
9995   19           Private   63363  Some-college             10   
9996   53           Private   58535       HS-grad              9   
9997   30           Private  342709       HS-grad              9   
9998   41  Self-emp-not-inc  134724     Assoc-voc             11   
9999   21           Private  252253  Some-college             10   

          marital-status         occupation   relationship   race     sex  \
0     Married-civ-spouse    Exec-managerial           Wife  White  Female   
1     Married-civ-spouse     

This function compares the values for each row in the columns 'age and 'hours-per-week' in the task_dataset and appends a new column with the comparison results. The comparison_result indicates True if value in each row in the first column is greater than the value in the second column an example is row 9996 having value for 'age as 53 and hours per week 46'.  and false if first column is lesser than the value in the second column an example is row 1 having value for 'age as 36 and hours per week 45'

## Task 1.8: Write a function that returns the names of countries with maximum and minimum average ages in the task_dataset (Hint: you can use the numpy module)

In [75]:
############# WRITE THE CODE IN THIS CELL####################

def find_extreme_avg_age_countries(task_dataset):
    
    # Calculate average age for each country
    avg_age_by_country = task_dataset.groupby('native-country')['age'].mean() 
    
    # Find the country with the maximum average age
    max_avg_age_country = avg_age_by_country.idxmax() 
    
    # Find the country with the minimum average age
    min_avg_age_country = avg_age_by_country.idxmin() 

    return max_avg_age_country, min_avg_age_country

max_age_country, min_age_country = find_extreme_avg_age_countries(task_dataset)
print(f"Country with maximum average age: {max_age_country}")
print(f"Country with minimum average age: {min_age_country}")


Country with maximum average age: Hungary
Country with minimum average age: Laos


The average age for each country is computed and the countries with the highest and lowest average ages identified

In [76]:
!pip install --upgrade jupyter_core


Word count: 319